lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/SpellChecker.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.analysis.hunspell;
+
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * A spell checker based on Hunspell dictionaries. Thread-safe. Not all Hunspell features are
+ * supported yet.
+ */
+public class SpellChecker {
+  private final Dictionary dictionary;
+  private final ThreadLocal<Stemmer> stemmer;
```

Comment:
A suggestion. If it's not absolutely necessary to make it thread-safe then I'd just use non-thread safe fields. If something is concurrent up the stack then it can be handled there (as a thread local SpellChecker). Or just store the dictionary and create an instance of SpellChecker on demand, it's a cheap operation?

Reply:
I did some performance testing and it seems that allocating Stemmer/BytesRef on each query seems fine indeed. Thanks!

Reply:
No problem at all. If we can postpone such optimizations, it's typically better solved in the particular context where it's used.